File: backend/src/main/java/es/udc/paproject/backend/model/services/PurchaseService.java
package es.udc.paproject.backend.model.services;

import com.stripe.exception.StripeException;
import es.udc.paproject.backend.model.entities.Purchase;
import es.udc.paproject.backend.model.entities.ShoppingCart;
import es.udc.paproject.backend.model.exceptions.*;

public interface PurchaseService {

    ShoppingCart addToCart(Long userId, Long shoppingCartId, Long productId, int quantity)
        throws InstanceNotFoundException, PermissionException, MaxQuantityExceededException, MaxItemsExceededException;

    ShoppingCart updateCartItem(Long userId, Long shoppingCartId, Long productId, int quantity)
            throws InstanceNotFoundException, PermissionException, MaxQuantityExceededException, MaxItemsExceededException;

    ShoppingCart removeCartItem(Long userId, Long shoppingCartId, Long productId)
        throws InstanceNotFoundException, PermissionException;

    Purchase createPurchase(Long userId, Long shoppingCartId, String postalAddress, String locality,
                            String region, String country, String postalCode)
            throws InstanceNotFoundException, PermissionException, EmptyShoppingCartException, StripeException, MaxItemsExceededException;

    void processPaymentForPurchase(String paymentMethodId, Purchase purchase) throws StripeException, PaymentProcessingException, InstanceNotFoundException, PaymentAlreadyProcessedException;

    Purchase findPurchaseById(Long userId, Long orderId) throws InstanceNotFoundException, PermissionException;

    Block<Purchase> findPurchases(Long userId, int page, int size);

}

File: frontend/src/backend/purchaseService.js
import {config, appFetch} from './appFetch';

export const addItemToCart = (shoppingCartId, productId, quantity, onSuccess, onErrors) =>
    appFetch(`/shopping/carts/${shoppingCartId}/addItem`,
        config('POST', {productId, quantity}), onSuccess, onErrors);


export const updateCartItem = (shoppingCartId, productId, quantity, onSuccess, onErrors) =>
    appFetch(`/shopping/carts/${shoppingCartId}/updateItem`,
        config('POST', {productId, quantity}), onSuccess, onErrors);

export const removeCartItem = (shoppingCartId, productId, onSuccess, onErrors) =>
    appFetch(`/shopping/carts/${shoppingCartId}/removeItem`,
        config('POST', {productId}), onSuccess, onErrors);

export const purchaseCart = (shoppingCartId, postalAddress, locality, region, country, postalCode, onSuccess, onErrors) =>
    appFetch(`/shopping/carts/${shoppingCartId}/purchase`,
        config('POST', {postalAddress, locality, region, country, postalCode}), onSuccess, onErrors);

export const findPurchase = (purchaseId, onSuccess) =>
    appFetch(`/shopping/purchases/${purchaseId}`, config('GET'), onSuccess);

export const findPurchases = ({page}, onSuccess) =>
    appFetch(`/shopping/purchases?page=${page}`, config('GET'), onSuccess);